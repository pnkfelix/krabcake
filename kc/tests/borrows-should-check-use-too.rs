use test_dependencies::{kc_borrow_mut, kc_as_raw, print_stack_of, print_tag_of};

pub fn main() {
    let mut val: u8 = 101;
    let l = std::ptr::addr_of!(val);
    print_stack_of(b"l_0\0".as_ptr(), l);
    let x = kc_borrow_mut!(val); // x = &mut val;
    print_stack_of(b"l_1\0".as_ptr(), l);
    let x_alias = kc_as_raw!(x, u8);
    print_stack_of(b"l_2\0".as_ptr(), l);
    let y = unsafe { kc_borrow_mut!(*x_alias) };
    print_tag_of(b"y\0".as_ptr(), y);
    print_stack_of(b"l_3\0".as_ptr(), l);
    let z = unsafe { kc_borrow_mut!(*x_alias) };
    print_tag_of(b"z\0".as_ptr(), z);
    print_stack_of(b"l_4\0".as_ptr(), l);
    println!("about to write via y");

    *y = 105;

    print_stack_of(b"l_5\0".as_ptr(), l);
    let end = *y;
    print_stack_of(b"l_6\0".as_ptr(), l);
    println!("Goodbye world, end: {}!", end);
    print_stack_of(b"l_7\0".as_ptr(), l);
}
