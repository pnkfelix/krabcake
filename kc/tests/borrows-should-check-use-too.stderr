lib.rs: normalizing output...
Hello world (from `rs_hello/src/lib.rs`)!
kc_main.c: dispatching code 4b43000a
print_stack_of `l_0` (0x10000000), no stack found
kc_main.c: dispatching code 4b430000
lib.rs: handle client request BORROW_MUT. borrowing_addr: 0x10000000
kc_main.c: dispatching code 4b43000a
print_stack_of `l_1` (0x10000000): [Unique(1)]
kc_main.c: dispatching code 4b430002
lib.rs: handle client request AS_RAW. borrowing_addr: 0x10000000
kc_main.c: dispatching code 4b43000a
print_stack_of `l_2` (0x10000000): [Unique(1), SharedRW]
kc_main.c: dispatching code 4b430000
lib.rs: handle client request BORROW_MUT. borrowing_addr: 0x10000000
kc_main.c: dispatching code 4b430009
print_tag_of `y` (0x10000000): 2
kc_main.c: dispatching code 4b43000a
print_stack_of `l_3` (0x10000000): [Unique(1), SharedRW, Unique(2)]
kc_main.c: dispatching code 4b430000
lib.rs: handle client request BORROW_MUT. borrowing_addr: 0x10000000
kc_main.c: dispatching code 4b430009
print_tag_of `z` (0x10000000): 3
kc_main.c: dispatching code 4b43000a
print_stack_of `l_4` (0x10000000): [Unique(1), SharedRW, Unique(3)]
ALERT could not find tag in stack for address 0x10000000 even though we are accessing it via pointer with tag 2
kc_main.c: dispatching code 4b43000a
print_stack_of `l_5` (0x10000000): []
ALERT could not find tag in stack for address 0x10000000 even though we are accessing it via pointer with tag 2
kc_main.c: dispatching code 4b43000a
print_stack_of `l_6` (0x10000000): []
kc_main.c: dispatching code 4b43000a
print_stack_of `l_7` (0x10000000): []
